ignite/src/main/java/org/hibernate/ogm/datastore/ignite/impl/IgniteDatastoreProvider.java

```diff
@@ -0,0 +1,398 @@
+		this.jdbcServices = serviceRegistryImplementor.getService( JdbcServices.class );
+	}
+
+	@Override
+	public Class<? extends SchemaDefiner> getSchemaDefinerType() {
+		return IgniteCacheInitializer.class;
+	}
+
+	public IgniteAtomicSequence atomicSequence(String name, int initialValue, boolean create) {
+		return cacheManager.atomicSequence( name, initialValue, create );
+	}
+
+	public boolean isClientMode() {
+		return cacheManager.configuration().isClientMode();
+	}
+
+	@Override
+	public Class<? extends QueryParserService> getDefaultQueryParserServiceType() {
+		return IgniteQueryParserService.class;
+	}
+
+	public String getGridName() {
+		return gridName;
+	}
+
+	@Override
+	public Class<? extends GridDialect> getDefaultDialect() {
+		return IgniteDialect.class;
+	}
+
+	public <T> List<T> affinityCall(String cacheName, Object affinityKey, SqlFieldsQuery query) {
+		ComputeForLocalQueries<T> call = new ComputeForLocalQueries<>( cacheName, query );
+		return cacheManager.compute().affinityCall( cacheName, affinityKey, call );
+	}
+
+	/**
+	 * this task runs inside server node and we shouldn't use any Hibernate classes here.
+	 */
+	private static class ComputeForLocalQueries<T> implements IgniteCallable<List<T>> {
+
+		private final String cacheName;
+		private final SqlFieldsQuery query;
+		@IgniteInstanceResource
+		private Ignite ignite;
+
+		public ComputeForLocalQueries(String cacheName, SqlFieldsQuery query) {
+			this.cacheName = cacheName;
+			this.query = query.setLocal( true );
+		}
+
+		@SuppressWarnings("unchecked")
+		@Override
+		public List<T> call() throws Exception {
+			IgniteCache<Object, BinaryObject> cache = ignite.cache( cacheName );
+			if ( cache == null ) {
+				throw new IgniteException( "Cache '" + cacheName + "' not found" );
+			}
+			cache = cache.withKeepBinary();
+			return (List<T>) cache.query( query ).getAll();
+		}
+	}
+
+	public SqlFieldsQuery createSqlFieldsQueryWithLog(String sql, QueryHints hints, Object... args) {
+		String comment = hints != null ? hints.toComment() : "";
+		jdbcServices.getSqlStatementLogger().logStatement( comment + sql );
+
+		SqlFieldsQuery query = new SqlFieldsQuery( sql );
+		if ( args != null ) {
+			query.setArgs( args );
+		}
+
+		return query;
+	}
+
+	/**
+	 * Converting entity key to cache key
+	 *
+	 * @param key entity key
+	 * @return string key
+	 */
+	public Object createKeyObject(EntityKey key) {
+		Object result = null;
+		if ( key.getColumnValues().length == 1 ) {
+			result = key.getColumnValues()[0];
+		}
+		else {
+			BinaryObjectBuilder builder = createBinaryObjectBuilder( findKeyType( key.getMetadata() ) );
+			for ( int i = 0; i < key.getColumnNames().length; i++ ) {
+				builder.setField( StringHelper.stringAfterPoint( key.getColumnNames()[i] ), key.getColumnValues()[i] );
+			}
+			result = builder.build();
+		}
+		return result;
+	}
+
+	public Object createAssociationKeyObject( RowKey rowKey, AssociationKeyMetadata keyMetadata ) {
```

Comment:
I think there is something wrong in the creation of the association key, it seems that you are not considering the entitykey and you risk to insert an element once even if it connected to two different entities.

I think this is the reason ManyToManyExtraTest fails, the student `mario` is inserted in the class math and english classes but when you count the element in the associations the end result is 3 because the association is only recorded for one of the classes. @Z-z-z-z Can you investaigate this?